File: genesis_audio_forge.py
# -*- coding: utf-8 -*-
"""Deterministic score planner for JANUS Genesis Audio Forge v1.

The planner is deliberately renderer-agnostic. It turns a bounded declarative
recipe + explicit world state + seed into a reproducible event plan. It does
not open audio devices, inspect the user, access the network, mutate gameplay,
or execute recipe-supplied code.
"""
from __future__ import annotations

import hashlib
import json
import math
from collections.abc import Mapping
from typing import Any

SCHEMA = "janus.genesis.audio_recipe.v1"
GENERATOR = "GENESIS_AUDIO_FORGE_V1"
VERSION = "1.0.0"
STEPS_PER_BAR = 16
ALLOWED_WAVES = frozenset({"sine", "triangle", "square", "sawtooth"})
ALLOWED_WORLD_KEYS = frozenset({"entropy", "depth", "portal_energy", "danger", "weather_intensity"})
DEFAULT_PROGRESSION = (0, 4, 5, 3)
BASE_PULSE_STEPS = (0, 4, 8, 12)
BASE_BASS_STEPS = (0, 3, 6, 8, 11, 14)
STAR_STEPS = (5, 13)


class AudioForgeViolation(ValueError):
    """Raised when a recipe or world-state input violates Audio Forge laws."""


def _clamp(value: Any, minimum: float, maximum: float) -> float:
    try:
        number = float(value)
    except (TypeError, ValueError) as exc:
        raise AudioForgeViolation(f"Expected numeric value, got {value!r}") from exc
    if not math.isfinite(number):
        raise AudioForgeViolation("Non-finite audio parameter is forbidden")
    return max(minimum, min(maximum, number))


def _stable(value: Any) -> Any:
    if isinstance(value, Mapping):
        return {str(key): _stable(value[key]) for key in sorted(value, key=str)}
    if isinstance(value, (list, tuple)):
        return [_stable(item) for item in value]
    return value


def canonical_bytes(value: Any) -> bytes:
    return json.dumps(_stable(value), ensure_ascii=False, separators=(",", ":")).encode("utf-8")


def sha256_hex(value: Any) -> str:
    return hashlib.sha256(canonical_bytes(value)).hexdigest()


def hash32(value: int) -> int:
    value &= 0xFFFFFFFF
    value = (value + 0x7ED55D16 + ((value << 12) & 0xFFFFFFFF)) & 0xFFFFFFFF
    value = (value ^ 0xC761C23C ^ (value >> 19)) & 0xFFFFFFFF
    value = (value + 0x165667B1 + ((value << 5) & 0xFFFFFFFF)) & 0xFFFFFFFF
    value = (value + 0xD3A2646C) & 0xFFFFFFFF
    value = (value ^ ((value << 9) & 0xFFFFFFFF)) & 0xFFFFFFFF
    value = (value + 0xFD7046C5 + ((value << 3) & 0xFFFFFFFF)) & 0xFFFFFFFF
    value = (value ^ 0xB55A4F09 ^ (value >> 16)) & 0xFFFFFFFF
    return value


def unit_rand(seed: int, bar: int, step: int, salt: int = 0) -> float:
    mixed = (
        (int(seed) & 0xFFFFFFFF)
        ^ (((bar + 1) * 0x9E3779B1) & 0xFFFFFFFF)
        ^ (((step + 1) * 0x85EBCA6B) & 0xFFFFFFFF)
        ^ (int(salt) & 0xFFFFFFFF)
    )
    return hash32(mixed) / 0xFFFFFFFF


def midi_to_hz(midi_note: int) -> float:
    return 440.0 * math.pow(2.0, (float(midi_note) - 69.0) / 12.0)


def _validate_int_list(value: Any, *, name: str, minimum: int, maximum: int, max_items: int) -> tuple[int, ...]:
    if not isinstance(value, (list, tuple)) or not value or len(value) > max_items:
        raise AudioForgeViolation(f"{name} must be a non-empty bounded array")
    out: list[int] = []
    for raw in value:
        if isinstance(raw, bool) or not isinstance(raw, (int, float)) or int(raw) != raw:
            raise AudioForgeViolation(f"{name} entries must be integers")
        number = int(raw)
        if number < minimum or number > maximum:
            raise AudioForgeViolation(f"{name} entry out of range: {number}")
        out.append(number)
    return tuple(out)


def validate_recipe(recipe: Mapping[str, Any]) -> dict[str, Any]:
    if not isinstance(recipe, Mapping):
        raise AudioForgeViolation("Audio recipe must be an object")
    if recipe.get("schema") != SCHEMA:
        raise AudioForgeViolation("Unsupported audio recipe schema")
    if recipe.get("generator") != GENERATOR:
        raise AudioForgeViolation("Unsupported audio generator")

    recipe_id = recipe.get("recipe_id")
    if not isinstance(recipe_id, str) or not recipe_id.startswith("genesis://audio/") or len(recipe_id) > 180:
        raise AudioForgeViolation("Invalid recipe_id")

    root_midi = int(_clamp(recipe.get("root_midi", 50), 24, 84))
    tempo_bpm = _clamp(recipe.get("tempo_bpm", 66), 36, 160)
    scale = _validate_int_list(recipe.get("scale"), name="scale", minimum=-24, maximum=24, max_items=16)
    motif = _validate_int_list(recipe.get("motif"), name="motif", minimum=-32, maximum=32, max_items=32)

    layers = recipe.get("layers")
    if not isinstance(layers, Mapping):
        raise AudioForgeViolation("layers must be an object")
    normalized_layers: dict[str, dict[str, Any]] = {}
    for name in ("pulse", "bass", "arp", "pad", "bells", "drone", "noise"):
        raw = layers.get(name, {})
        if not isinstance(raw, Mapping):
            raise AudioForgeViolation(f"layers.{name} must be an object")
        wave = str(raw.get("wave", "sine"))
        if wave not in ALLOWED_WAVES:
            raise AudioForgeViolation(f"Unsupported oscillator wave: {wave}")
        normalized_layers[name] = {
            "enabled": bool(raw.get("enabled", True)),
            "density": _clamp(raw.get("density", 0.5), 0.0, 1.0),
            "gain": _clamp(raw.get("gain", 0.03), 0.0001, 0.2),
            "wave": wave,
        }

    bindings = recipe.get("world_bindings", {})
    if not isinstance(bindings, Mapping):
        raise AudioForgeViolation("world_bindings must be an object")
    unknown_bindings = sorted(set(bindings) - ALLOWED_WORLD_KEYS)
    if unknown_bindings:
        raise AudioForgeViolation(f"Unsupported world bindings: {', '.join(unknown_bindings)}")

    return {
        "schema": SCHEMA,
        "generator": GENERATOR,
        "generator_version": str(recipe.get("generator_version", VERSION)),
        "recipe_id": recipe_id,
        "tempo_bpm": tempo_bpm,
        "root_midi": root_midi,
        "scale": scale,
        "motif": motif,
        "layers": normalized_layers,
        "world_bindings": {str(k): str(v) for k, v in bindings.items()},
    }


def normalize_world_state(world_state: Mapping[str, Any] | None) -> dict[str, float]:
    raw = world_state or {}
    if not isinstance(raw, Mapping):
        raise AudioForgeViolation("world_state must be an object")
    unknown = sorted(set(raw) - ALLOWED_WORLD_KEYS)
    if unknown:
        raise AudioForgeViolation(f"Hidden or unsupported world telemetry is forbidden: {', '.join(unknown)}")
    return {key: _clamp(raw.get(key, 0.0), 0.0, 1.0) for key in sorted(ALLOWED_WORLD_KEYS)}


def _mode_midi(recipe: Mapping[str, Any], degree: int, octave: int = 0) -> int:
    scale = tuple(recipe["scale"])
    length = len(scale)
    octave_shift, index = divmod(int(degree), length)
    return int(recipe["root_midi"]) + int(scale[index]) + 12 * (octave + octave_shift)


def _event(kind: str, step: int, **payload: Any) -> dict[str, Any]:
    return {"kind": kind, "step": int(step), **payload}


def plan_bar(
    recipe: Mapping[str, Any],
    *,
    seed: int,
    bar_index: int,
    world_state: Mapping[str, Any] | None = None,
) -> dict[str, Any]:
    """Create a deterministic, renderer-independent 16-step score plan."""
    if isinstance(seed, bool) or not isinstance(seed, int) or seed < 0 or seed > 0xFFFFFFFF:
        raise AudioForgeViolation("seed must be an unsigned 32-bit integer")
    if isinstance(bar_index, bool) or not isinstance(bar_index, int) or bar_index < 0:
        raise AudioForgeViolation("bar_index must be a non-negative integer")

    normalized = validate_recipe(recipe)
    world = normalize_world_state(world_state)
    layers = normalized["layers"]
    progression_root = DEFAULT_PROGRESSION[bar_index % len(DEFAULT_PROGRESSION)]

    entropy = world["entropy"]
    depth = world["depth"]
    portal = world["portal_energy"]
    danger = world["danger"]
    weather = world["weather_intensity"]
    tempo = _clamp(normalized["tempo_bpm"] + portal * 18.0 + danger * 10.0 - depth * 5.0, 36, 180)
    filter_hz = _clamp(2200.0 + entropy * 2400.0 + portal * 1300.0 - depth * 600.0, 500, 7600)
    register_shift = -1 if depth > 0.70 else (1 if portal > 0.82 else 0)

    events: list[dict[str, Any]] = []
    if layers["drone"]["enabled"]:
        events.append(_event("drone", 0, midi=_mode_midi(normalized, progression_root, -2 + register_shift), gain=layers["drone"]["gain"], wave=layers["drone"]["wave"]))

    if layers["pad"]["enabled"]:
        chord = [_mode_midi(normalized, progression_root + degree, -1 + register_shift) for degree in (0, 2, 4)]
        events.append(_event("pad", 0, midi=chord, gain=layers["pad"]["gain"], wave=layers["pad"]["wave"]))

    motif = tuple(normalized["motif"])
    for step in range(STEPS_PER_BAR):
        if layers["pulse"]["enabled"]:
            threshold = min(1.0, layers["pulse"]["density"] + danger * 0.18 + portal * 0.12)
            if step in BASE_PULSE_STEPS or unit_rand(seed, bar_index, step, 101) < threshold * 0.22:
                events.append(_event("pulse", step, gain=layers["pulse"]["gain"] * (1.0 + danger * 0.25), wave=layers["pulse"]["wave"]))

        if layers["bass"]["enabled"] and step in BASE_BASS_STEPS:
            threshold = min(1.0, layers["bass"]["density"] + danger * 0.15)
            if unit_rand(seed, bar_index, step, 202) < threshold:
                degree = progression_root + (0, 0, 2, 0, 4, 2)[BASE_BASS_STEPS.index(step)]
                events.append(_event("bass", step, midi=_mode_midi(normalized, degree, -2 + register_shift), gain=layers["bass"]["gain"], wave=layers["bass"]["wave"]))

        if layers["arp"]["enabled"] and step % 2 == 1:
            threshold = min(1.0, layers["arp"]["density"] + portal * 0.22 + entropy * 0.10)
            if unit_rand(seed, bar_index, step, 303) < threshold:
                degree = progression_root + motif[(step // 2 + bar_index) % len(motif)]
                events.append(_event("arp", step, midi=_mode_midi(normalized, degree, register_shift), gain=layers["arp"]["gain"], wave=layers["arp"]["wave"], pan=round(unit_rand(seed, bar_index, step, 304) * 1.4 - 0.7, 6)))

        if layers["bells"]["enabled"] and step in STAR_STEPS:
            threshold = min(1.0, layers["bells"]["density"] + portal * 0.25)
            if unit_rand(seed, bar_index, step, 404) < threshold:
                degree = progression_root + motif[(step + bar_index) % len(motif)] + 4
                events.append(_event("bell", step, midi=_mode_midi(normalized, degree, 1 + register_shift), gain=layers["bells"]["gain"], wave=layers["bells"]["wave"], pan=(-0.55 if step < 8 else 0.55)))

        if layers["noise"]["enabled"]:
            threshold = min(1.0, layers["noise"]["density"] * 0.25 + weather * 0.42 + entropy * 0.08)
            if unit_rand(seed, bar_index, step, 505) < threshold:
                events.append(_event("noise", step, gain=layers["noise"]["gain"] * (0.5 + weather * 0.8), cutoff_hz=round(filter_hz + weather * 900.0, 3)))

    plan: dict[str, Any] = {
        "schema": "janus.genesis.audio_score_plan.v1",
        "generator": GENERATOR,
        "generator_version": VERSION,
        "recipe_id": normalized["recipe_id"],
        "recipe_sha256": sha256_hex(normalized),
        "seed": seed,
        "bar_index": bar_index,
        "steps_per_bar": STEPS_PER_BAR,
        "tempo_bpm": round(tempo, 6),
        "filter_hz": round(filter_hz, 3),
        "world_state": world,
        "events": events,
        "authority": {
            "presentation_only": True,
            "world_mutation": False,
            "hidden_human_telemetry": False,
            "network_access": False,
            "arbitrary_code_execution": False,
        },
    }
    plan["plan_sha256"] = sha256_hex(plan)
    return plan


LAWS = (
    "AUDIO_RECIPE_NE_ARBITRARY_CODE",
    "GENERATED_AUDIO_NE_STORED_TRACK",
    "SOUND_PRESENTATION_NE_WORLD_AUTHORITY",
    "WORLD_STATE_TO_AUDIO_ALLOWED",
    "AUDIO_TO_WORLD_MUTATION_DEFAULT_DENY",
    "HIDDEN_HUMAN_TELEMETRY_FORBIDDEN",
    "USER_GESTURE_REQUIRED_FOR_BROWSER_AUDIO",
    "SAME_RECIPE_SEED_VERSION_EQ_SAME_SCORE_PLAN",
    "RENDERER_NE_RECIPE",
    "PROVENANCE_SURVIVES_EXTRACTION",
)

File: genesis-audio-forge.js
(() => {
  'use strict';

  const VERSION = '1.0.0';
  const GENERATOR = 'GENESIS_AUDIO_FORGE_V1';
  const STEPS = 16;
  const ALLOWED_WAVES = new Set(['sine', 'triangle', 'square', 'sawtooth']);
  const PROGRESSION = [0, 4, 5, 3];
  const PULSE_STEPS = new Set([0, 4, 8, 12]);
  const BASS_STEPS = [0, 3, 6, 8, 11, 14];
  const STAR_STEPS = new Set([5, 13]);

  const DEFAULT_RECIPE = Object.freeze({
    schema: 'janus.genesis.audio_recipe.v1',
    generator: GENERATOR,
    generator_version: VERSION,
    recipe_id: 'genesis://audio/biome/liminal-hall/v1',
    tempo_bpm: 66,
    root_midi: 50,
    scale: [0, 2, 4, 6, 7, 9, 11],
    motif: [0, 4, 2, 5, 1, 6, 4, 2],
    layers: {
      pulse: {enabled: true, density: .42, gain: .030, wave: 'sine'},
      bass: {enabled: true, density: .58, gain: .045, wave: 'triangle'},
      arp: {enabled: true, density: .70, gain: .024, wave: 'triangle'},
      pad: {enabled: true, density: 1, gain: .020, wave: 'sine'},
      bells: {enabled: true, density: .38, gain: .018, wave: 'sine'},
      drone: {enabled: true, density: 1, gain: .012, wave: 'sine'},
      noise: {enabled: true, density: .22, gain: .009, wave: 'sine'}
    }
  });

  const state = {
    enabled: false, running: false, ctx: null, master: null, filter: null,
    delay: null, feedback: null, compressor: null, noiseBuffer: null,
    timer: 0, nextStepTime: 0, step: 0, bar: 0, seed: 0,
    recipe: clone(DEFAULT_RECIPE),
    world: {entropy: 0, depth: 0, portal_energy: 0, danger: 0, weather_intensity: 0}
  };

  function clone(value) { return JSON.parse(JSON.stringify(value)); }
  function clamp(value, min, max) {
    const n = Number(value);
    if (!Number.isFinite(n)) throw new TypeError('GENESIS Audio Forge expects finite numeric parameters');
    return Math.max(min, Math.min(max, n));
  }
  function midiToHz(midi) { return 440 * Math.pow(2, (Number(midi) - 69) / 12); }
  function hash32(x) {
    x |= 0; x = (x + 0x7ed55d16 + (x << 12)) | 0;
    x = (x ^ 0xc761c23c ^ (x >>> 19)) | 0;
    x = (x + 0x165667b1 + (x << 5)) | 0;
    x = (x + 0xd3a2646c) | 0; x = (x ^ (x << 9)) | 0;
    x = (x + 0xfd7046c5 + (x << 3)) | 0;
    x = (x ^ 0xb55a4f09 ^ (x >>> 16)) | 0;
    return x >>> 0;
  }
  function unitRand(step, salt = 0) {
    const mixed = (state.seed ^ Math.imul(state.bar + 1, 0x9e3779b1) ^ Math.imul(step + 1, 0x85ebca6b) ^ salt) >>> 0;
    return hash32(mixed) / 0xffffffff;
  }
  function secureSeed() {
    const a = new Uint32Array(1);
    if (globalThis.crypto?.getRandomValues) globalThis.crypto.getRandomValues(a);
    else a[0] = (Date.now() * 2654435761) >>> 0;
    return a[0] >>> 0;
  }

  function normalizeLayer(raw, fallback) {
    const source = raw && typeof raw === 'object' ? raw : fallback;
    const wave = String(source.wave || fallback.wave);
    if (!ALLOWED_WAVES.has(wave)) throw new TypeError(`Unsupported oscillator wave: ${wave}`);
    return {
      enabled: source.enabled !== false,
      density: clamp(source.density ?? fallback.density, 0, 1),
      gain: clamp(source.gain ?? fallback.gain, .0001, .2),
      wave
    };
  }

  function validateRecipe(input) {
    if (!input || typeof input !== 'object') throw new TypeError('Audio recipe must be an object');
    if (input.schema !== 'janus.genesis.audio_recipe.v1') throw new TypeError('Unsupported audio recipe schema');
    if (input.generator !== GENERATOR) throw new TypeError('Unsupported audio generator');
    if (typeof input.recipe_id !== 'string' || !input.recipe_id.startsWith('genesis://audio/')) throw new TypeError('Invalid recipe_id');
    if (!Array.isArray(input.scale) || !input.scale.length || input.scale.length > 16) throw new TypeError('Invalid scale');
    if (!Array.isArray(input.motif) || !input.motif.length || input.motif.length > 32) throw new TypeError('Invalid motif');
    const layers = {};
    for (const name of Object.keys(DEFAULT_RECIPE.layers)) layers[name] = normalizeLayer(input.layers?.[name], DEFAULT_RECIPE.layers[name]);
    return {
      schema: input.schema,
      generator: GENERATOR,
      generator_version: String(input.generator_version || VERSION),
      recipe_id: input.recipe_id,
      tempo_bpm: clamp(input.tempo_bpm ?? 66, 36, 160),
      root_midi: Math.round(clamp(input.root_midi ?? 50, 24, 84)),
      scale: input.scale.map(v => Math.round(clamp(v, -24, 24))),
      motif: input.motif.map(v => Math.round(clamp(v, -32, 32))),
      layers
    };
  }

  function setWorldState(input = {}) {
    const allowed = new Set(['entropy', 'depth', 'portal_energy', 'danger', 'weather_intensity']);
    for (const key of Object.keys(input)) if (!allowed.has(key)) throw new TypeError(`Unsupported/hidden telemetry field: ${key}`);
    for (const key of allowed) state.world[key] = clamp(input[key] ?? 0, 0, 1);
    return getState();
  }

  function impulse(ctx, seconds = 2.2, decay = 2.6) {
    const length = Math.floor(ctx.sampleRate * seconds);
    const buffer = ctx.createBuffer(2, length, ctx.sampleRate);
    for (let ch = 0; ch < 2; ch++) {
      const data = buffer.getChannelData(ch);
      for (let i = 0; i < length; i++) {
        const env = Math.pow(1 - i / length, decay);
        const pseudo = (((i * 1103515245 + ch * 12345) >>> 8) & 0xffff) / 32767.5 - 1;
        data[i] = pseudo * env * .34;
      }
    }
    return buffer;
  }

  function buildNoiseBuffer(ctx) {
    const length = Math.floor(ctx.sampleRate * .7);
    const buffer = ctx.createBuffer(1, length, ctx.sampleRate);
    const data = buffer.getChannelData(0);
    let seed = state.seed || 0x9e3779b9;
    for (let i = 0; i < length; i++) {
      seed = (Math.imul(seed, 1664525) + 1013904223) >>> 0;
      data[i] = seed / 0xffffffff * 2 - 1;
    }
    return buffer;
  }

  function ensureGraph() {
    if (state.ctx) return state.ctx;
    const Ctx = globalThis.AudioContext || globalThis.webkitAudioContext;
    if (!Ctx) return null;
    const ctx = new Ctx();
    const master = ctx.createGain();
    const filter = ctx.createBiquadFilter();
    const delay = ctx.createDelay(1.2);
    const feedback = ctx.createGain();
    const reverb = ctx.createConvolver();
    const compressor = ctx.createDynamicsCompressor();
    master.gain.value = .0001;
    filter.type = 'lowpass'; filter.frequency.value = 3200; filter.Q.value = .55;
    delay.delayTime.value = .285; feedback.gain.value = .20;
    reverb.buffer = impulse(ctx);
    compressor.threshold.value = -18; compressor.knee.value = 18; compressor.ratio.value = 3;
    compressor.attack.value = .008; compressor.release.value = .36;
    filter.connect(master);
    filter.connect(delay); delay.connect(feedback); feedback.connect(delay); delay.connect(master);
    filter.connect(reverb); reverb.connect(master);
    master.connect(compressor).connect(ctx.destination);
    Object.assign(state, {ctx, master, filter, delay, feedback, compressor, noiseBuffer: buildNoiseBuffer(ctx)});
    return ctx;
  }

  function pan(node, value = 0) {
    if (!state.ctx?.createStereoPanner) return node;
    const p = state.ctx.createStereoPanner(); p.pan.value = clamp(value, -1, 1); node.connect(p); return p;
  }

  function voice({freq, when, duration = .2, gain = .02, wave = 'sine', cutoff = 3200, panValue = 0, detune = 0, sweepTo = 0}) {
    if (!state.enabled || !state.ctx || !state.filter) return;
    const osc = state.ctx.createOscillator();
    const amp = state.ctx.createGain();
    const tone = state.ctx.createBiquadFilter();
    osc.type = ALLOWED_WAVES.has(wave) ? wave : 'sine';
    osc.frequency.setValueAtTime(freq, when);
    if (sweepTo > 0) osc.frequency.exponentialRampToValueAtTime(sweepTo, when + duration * .82);
    osc.detune.setValueAtTime(detune, when);
    tone.type = 'lowpass'; tone.frequency.setValueAtTime(cutoff, when); tone.Q.value = .65;
    amp.gain.setValueAtTime(.0001, when);
    amp.gain.exponentialRampToValueAtTime(Math.max(.0002, gain), when + .018);
    amp.gain.exponentialRampToValueAtTime(.0001, when + duration);
    osc.connect(tone).connect(amp);
    pan(amp, panValue).connect(state.filter);
    osc.start(when); osc.stop(when + duration + .04);
  }

  function noiseHit(when, gain, cutoff) {
    if (!state.enabled || !state.ctx || !state.noiseBuffer || !state.master) return;
    const src = state.ctx.createBufferSource();
    const hp = state.ctx.createBiquadFilter();
    const amp = state.ctx.createGain();
    src.buffer = state.noiseBuffer; hp.type = 'highpass'; hp.frequency.setValueAtTime(cutoff, when);
    amp.gain.setValueAtTime(Math.max(.0002, gain), when); amp.gain.exponentialRampToValueAtTime(.0001, when + .05);
    src.connect(hp).connect(amp).connect(state.master);
    src.start(when, Math.min(.3, unitRand(state.step, 991) * .3)); src.stop(when + .07);
  }

  function modeMidi(degree, octave = 0) {
    const scale = state.recipe.scale;
    const len = scale.length;
    const octShift = Math.floor(degree / len);
    const index = ((degree % len) + len) % len;
    return state.recipe.root_midi + scale[index] + 12 * (octave + octShift);
  }

  function effectiveBpm() {
    const w = state.world;
    return clamp(state.recipe.tempo_bpm + w.portal_energy * 18 + w.danger * 10 - w.depth * 5, 36, 180);
  }
  function stepSeconds() { return (60 / effectiveBpm()) / 4; }

  function scheduleStep(when) {
    const step = state.step;
    const root = PROGRESSION[state.bar % PROGRESSION.length];
    const w = state.world;
    const layers = state.recipe.layers;
    const register = w.depth > .70 ? -1 : (w.portal_energy > .82 ? 1 : 0);
    const filterHz = clamp(2200 + w.entropy * 2400 + w.portal_energy * 1300 - w.depth * 600, 500, 7600);
    state.filter?.frequency.setTargetAtTime(filterHz, when, .05);

    if (step === 0 && layers.drone.enabled) voice({freq: midiToHz(modeMidi(root, -2 + register)), when, duration: stepSeconds() * 15, gain: layers.drone.gain, wave: layers.drone.wave, cutoff: 900});
    if (step === 0 && layers.pad.enabled) [0, 2, 4].forEach((d, i) => voice({freq: midiToHz(modeMidi(root + d, -1 + register)), when, duration: stepSeconds() * 15.2, gain: layers.pad.gain, wave: layers.pad.wave, cutoff: 1900, panValue: (i - 1) * .35, detune: i % 2 ? 6 : -6}));

    const pulseThreshold = Math.min(1, layers.pulse.density + w.danger * .18 + w.portal_energy * .12);
    if (layers.pulse.enabled && (PULSE_STEPS.has(step) || unitRand(step, 101) < pulseThreshold * .22)) voice({freq: 86, sweepTo: 43, when, duration: .18, gain: layers.pulse.gain * (1 + w.danger * .25), wave: 'sine', cutoff: 700});

    if (layers.bass.enabled && BASS_STEPS.includes(step) && unitRand(step, 202) < Math.min(1, layers.bass.density + w.danger * .15)) {
      const degree = root + [0, 0, 2, 0, 4, 2][BASS_STEPS.indexOf(step)];
      voice({freq: midiToHz(modeMidi(degree, -2 + register)), when, duration: .24, gain: layers.bass.gain, wave: layers.bass.wave, cutoff: 760, panValue: step < 8 ? -.18 : .18});
    }

    if (layers.arp.enabled && step % 2 === 1 && unitRand(step, 303) < Math.min(1, layers.arp.density + w.portal_energy * .22 + w.entropy * .10)) {
      const degree = root + state.recipe.motif[(Math.floor(step / 2) + state.bar) % state.recipe.motif.length];
      voice({freq: midiToHz(modeMidi(degree, register)), when, duration: .18, gain: layers.arp.gain, wave: layers.arp.wave, cutoff: filterHz, panValue: unitRand(step, 304) * 1.4 - .7});
    }

    if (layers.bells.enabled && STAR_STEPS.has(step) && unitRand(step, 404) < Math.min(1, layers.bells.density + w.portal_energy * .25)) {
      const degree = root + state.recipe.motif[(step + state.bar) % state.recipe.motif.length] + 4;
      const freq = midiToHz(modeMidi(degree, 1 + register));
      voice({freq, when, duration: .72, gain: layers.bells.gain, wave: 'sine', cutoff: 6900, panValue: step < 8 ? -.55 : .55});
      voice({freq: freq * 2.01, when, duration: .42, gain: layers.bells.gain * .28, wave: 'sine', cutoff: 7600, panValue: step < 8 ? .35 : -.35});
    }

    if (layers.noise.enabled && unitRand(step, 505) < Math.min(1, layers.noise.density * .25 + w.weather_intensity * .42 + w.entropy * .08)) noiseHit(when, layers.noise.gain * (.5 + w.weather_intensity * .8), filterHz + w.weather_intensity * 900);
  }

  function scheduler() {
    if (!state.running || !state.enabled || !state.ctx) return;
    while (state.nextStepTime < state.ctx.currentTime + .14) {
      scheduleStep(state.nextStepTime);
      state.nextStepTime += stepSeconds();
      state.step += 1;
      if (state.step >= STEPS) { state.step = 0; state.bar += 1; }
    }
  }

  async function enable({seed = null, recipe = null, world_state = null} = {}) {
    if (recipe) state.recipe = validateRecipe(recipe);
    if (world_state) setWorldState(world_state);
    state.seed = seed == null ? secureSeed() : (Number(seed) >>> 0);
    const ctx = ensureGraph();
    if (!ctx) return false;
    await ctx.resume();
    state.enabled = true;
    state.master.gain.cancelScheduledValues(ctx.currentTime);
    state.master.gain.setTargetAtTime(.08, ctx.currentTime, .08);
    if (!state.running) {
      state.running = true; state.nextStepTime = ctx.currentTime + .04; state.step = 0; state.bar = 0;
      state.timer = globalThis.setInterval(scheduler, 25);
    }
    globalThis.dispatchEvent?.(new CustomEvent('genesis:audio-forge-state', {detail: getState()}));
    return true;
  }

  function disable() {
    state.enabled = false;
    if (state.master && state.ctx) state.master.gain.setTargetAtTime(.0001, state.ctx.currentTime, .06);
    return getState();
  }

  function stop() {
    state.running = false;
    if (state.timer) globalThis.clearInterval(state.timer);
    state.timer = 0;
    disable();
    return getState();
  }

  function configure(recipe) { state.recipe = validateRecipe(recipe); return getState(); }

  function cue(kind, intensity = .5) {
    if (!state.enabled || !state.ctx) return false;
    const level = clamp(intensity, 0, 1);
    const now = state.ctx.currentTime + .012;
    const root = state.recipe.root_midi;
    const recipes = {
      portal_open: [0, 7, 12, 19],
      discovery: [0, 4, 7, 14],
      danger: [0, 1, 6, 7],
      resolve: [12, 9, 5, 0]
    };
    const intervals = recipes[kind] || recipes.discovery;
    intervals.forEach((interval, i) => voice({freq: midiToHz(root + 12 + interval), when: now + i * .075, duration: .45 + level * .25, gain: .008 + level * .012, wave: i % 2 ? 'triangle' : 'sine', cutoff: 3500 + level * 3000, panValue: (i - 1.5) * .28}));
    return true;
  }

  function getState() {
    return {
      version: VERSION, generator: GENERATOR, enabled: state.enabled, running: state.running,
      seed: state.seed, recipe_id: state.recipe.recipe_id, bar: state.bar, step: state.step,
      world_state: {...state.world}, presentation_only: true, world_mutation: false,
      hidden_human_telemetry: false, network_access: false, arbitrary_code_execution: false,
      user_gesture_required: true
    };
  }

  if (typeof document !== 'undefined') document.addEventListener('visibilitychange', () => {
    if (!state.master || !state.ctx) return;
    state.master.gain.setTargetAtTime(document.hidden || !state.enabled ? .0001 : .08, state.ctx.currentTime, .08);
  });

  globalThis.GENESIS_AUDIO_FORGE = Object.freeze({version: VERSION, configure, setWorldState, enable, disable, stop, cue, getState});
  globalThis.dispatchEvent?.(new CustomEvent('genesis:audio-forge-ready', {detail: {version: VERSION, generator: GENERATOR, presentation_only: true}}));
})();
